servicetalk-examples/grpc/compression/src/main/java/io/servicetalk/examples/grpc/compression/CompressionExampleServer.java

```diff
@@ -0,0 +1,80 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.examples.grpc.compression;
+
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.encoding.api.ContentCodec;
+import io.servicetalk.encoding.api.ContentCodings;
+import io.servicetalk.grpc.api.GrpcServiceContext;
+import io.servicetalk.grpc.netty.GrpcServers;
+import io.servicetalk.http.api.ContentCodingHttpServiceFilter;
+
+import io.grpc.examples.compression.Greeter.GreeterService;
+import io.grpc.examples.compression.Greeter.ServiceFactory;
+import io.grpc.examples.compression.HelloReply;
+import io.grpc.examples.compression.HelloRequest;
+
+import java.util.List;
+
+import static io.servicetalk.concurrent.api.Single.succeeded;
+
+/**
+ * A simple extension of the gRPC "Hello World" example which demonstrates
+ * compression of the request and response bodies.
+ */
+public class CompressionExampleServer {
+
+    /**
+     * Supported encodings for the request. Requests using unsupported encodings will receive an HTTP 415
+     * "Unsupported Media Type" response.
+     */
+    private static final List<ContentCodec> SUPPORTED_REQ_ENCODINGS = List.of(
+            ContentCodings.gzipDefault(),
+            ContentCodings.deflateDefault(),
+            ContentCodings.identity()
+    );
+
+    /**
+     * Supported encodings for the response in preferred order. These will be matched against the list of encodings
+     * provided by the client to choose a mutually agreeable encoding.
+     */
+    private static final List<ContentCodec> SUPPORTED_RESP_ENCODINGS = List.of(
+            // For the purposes of this example we disable GZip for the response compression and use the client's second
+            // choice (deflate) to demonstrate that negotiation of compression algorithm is handled correctly.
+            /* ContentCodings.gzipDefault(), */
+            ContentCodings.deflateDefault(),
+            ContentCodings.identity()
+    );
+
+    public static void main(String... args) throws Exception {
+        GrpcServers.forPort(8080)
+                // Adds a content coding service filter that includes the encodings supported for requests and the
+                // preferred encodings for responses. Responses will automatically be compressed if the request includes
+                // a mutually agreeable compression encoding that the client indicates they will accept and that the
+                // server supports.
+                .appendHttpServiceFilter(new ContentCodingHttpServiceFilter(SUPPORTED_REQ_ENCODINGS, SUPPORTED_RESP_ENCODINGS))
```

Comment:
The `ContentCodingHttpServiceFilter` only applies to HTTP compression, not gRPC.
Look at https://github.com/apple/servicetalk/blob/main/servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/api/GrpcMessageEncodingTest.java#L420 for how to enable for a gRPC service. This API is code-generated.

Reply:
Also please update the comment in relation to the content-coding reference.

Reply:
Are both needed or just the change to use the other ServiceFactory overload?

Reply:
Only the service factory change. The filter is only for HTTP. You can verify behaviour by logging the frames, see. '.enableWireLogging()'

Reply:
Compression for gRPC and HTTP works at different levels:
- HTTP compresses the entire payload body and uses the `content-encoding` header to signal which compression was applied.
- gRPC compresses each proto-message independently and uses `Compressed-Flag` to mark which are compressed and `grpc-encoding` header to tell which algorithm was used.
 
Therefore, filters that are used for HTTP compression can not be used with gRPC because they break protocol: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

Reply:
OK, thank you for the explanation. I will remove the HTTP filter.